ehrql/query_engines/mssql.py

```diff
@@ -201,16 +191,48 @@ def execute_query_with_results(self, connection, query, query_id):
                 log=log.info,
             )
 
-            yield from fetch_table_in_batches(
-                execute_with_retry,
-                results_table,
-                key_column_index=0,
-                key_is_unique=key_is_unique,
-                # This value was copied from the previous cohortextractor. I suspect it
-                # has no real scientific basis.
-                batch_size=32000,
-                log=log.info,
-            )
+            if query_type is self.QueryType.PATIENT_LEVEL:
+                yield from self.fetch_results_batched(
+                    execute_with_retry,
+                    query,
+                    key_is_unique=True,
+                )
+            elif query_type is self.QueryType.EVENT_LEVEL:
+                yield from self.fetch_results_batched(
+                    execute_with_retry,
+                    query,
+                    key_is_unique=False,
+                )
+            elif query_type is self.QueryType.AGGREGATED:
+                yield from self.fetch_results_in_one_go(
+                    execute_with_retry,
+                    query,
+                )
+            else:
+                assert False, f"Unhandled query type: {query_type}"
+
+    def fetch_results_batched(self, execute, query, key_is_unique):
+        # We're expecting queries in a very specific form which is "select everything
+        # from one table"; so we assert that each query has this form and retrieve a
+        # reference to the table
+        results_table = query.get_final_froms()[0]
+        assert str(query) == str(sqlalchemy.select(results_table))
+
+        return fetch_table_in_batches(
+            execute,
+            results_table,
+            key_column_index=0,
```

Comment:
This was changed from explicitly using "patient_id" as the key column for measures - it could go back now. Or maybe just a comment that the column is generally expected to be patient_id?

Reply:
Ah yeah, you're right: it would be better to be explicit here. I'll add that back.